cuqi/experimental/mcmc/_sampler.py

```diff
@@ -0,0 +1,324 @@
+
+        """
+
+        # Initialize batch handler
+        if batch_size > 0:
+            batch_handler = _BatchHandler(batch_size, sample_path)
+
+        # Draw samples
+        for _ in progressbar( range(Ns) ):
+            
+            # Perform one step of the sampler
+            acc = self.step()
+
+            # Store samples
+            self._acc.append(acc)
+            self._samples.append(self.current_point)
+
+            # Add sample to batch
+            if batch_size > 0:
+                batch_handler.add_sample(self.current_point)
+
+            # Call callback function if specified            
+            self._call_callback(self.current_point, len(self._samples)-1)
+                
+        return self
+    
+
+    def warmup(self, Nb, tune_freq=0.1) -> 'SamplerNew':
+        """ Warmup the sampler by drawing Nb samples.
+
+        Parameters
+        ----------
+        Nb : int
+            The number of samples to draw during warmup.
+
+        tune_freq : float, optional
+            The frequency of tuning. Tuning is performed every tune_freq*Nb samples.
+
+        """
+
+        tune_interval = max(int(tune_freq * Nb), 1)
+
+        # Draw warmup samples with tuning
+        for idx in progressbar(range(Nb)):
+
+            # Perform one step of the sampler
+            acc = self.step()
+
+            # Tune the sampler at tuning intervals
+            if (idx + 1) % tune_interval == 0:
+                self.tune(tune_interval, idx // tune_interval) 
+
+            # Store samples
+            self._acc.append(acc)
+            self._samples.append(self.current_point)
+
+            # Call callback function if specified
+            self._call_callback(self.current_point, len(self._samples)-1)
+
+        return self
+
+    def _call_callback(self, sample, sample_index):
+        """ Calls the callback function. Assumes input is sample and sample index"""
+        if self.callback is not None:
+            self.callback(sample, sample_index)
+
+
+class ProposalBasedSamplerNew(SamplerNew, ABC):
+    """ Abstract base class for samplers that use a proposal distribution. """
+    def __init__(self, target, proposal=None, scale=1, **kwargs):
+        """ Initializer for proposal based samplers. 
+
+        Parameters
+        ----------
+        target : cuqi.density.Density
+            The target density.
+
+        proposal : cuqi.distribution.Distribution, optional
+            The proposal distribution. If not specified, the default proposal is used.
+
+        scale : float, optional
+            The scale parameter for the proposal distribution.
+
+        **kwargs : dict
+            Additional keyword arguments passed to the :class:`SamplerNew` initializer.
+
+        """
+
+        super().__init__(target, **kwargs)
+
+        self.current_point = self.initial_point
+        self.current_target = self.target.logd(self.current_point)
+        self.proposal = proposal
+        self.scale = scale
+
```

Comment:
I suggest adding self._scale_temp here. 
```
self.scale_temp = scale
```

Reply:
I would rather fix the actual bug and avoid this "hack" in the base class. I added it to #378